### PR TITLE
Add system.info event for agents

### PR DESCRIPTION
**Pull request by mjangda**
Sends an agent information about the current state of the system like active chats and connected operators.

The happychat implementation will set up a poller so we can periodically send updates.

See #25

### CODE REVIEW

File: src/agent.js [hunk 1]
```diff
@@ -20,10 +20,23 @@ const onAuthorized = ( { socket, events } ) => ( agent ) => {
...
+
 	socket.on( 'role.add', ( role, done ) => {
 		debug( 'agent joining role', role )
+		events.emit( 'role.add', role )
 		socket.join( role, done )
 	} )
+
+	socket.on( 'disconnect', () => {
+		for ( const room in socket.rooms ) {
+			events.emit( 'role.remove', room )
```
Comment (mjangda) [inline src/agent.js]: Added these events so `happychat` can listen for leave and remove the interval when the room is empty.

File: src/operator/index.js [hunk 2]
```diff
@@ -305,6 +305,8 @@ export default io => {
 
 	events.io = io
 
+	events.getIdentities = () => selectIdentities( store.getState() )
```
Comment (mjangda) [inline src/operator/index.js]: Not sure if there's a better place to expose this selector.

**Comment by mjangda** [comment]
@beaucollins can you take a look?

File: src/agent.js [hunk 3]
```diff
@@ -20,10 +20,17 @@ const onAuthorized = ( { socket, events } ) => ( agent ) => {
...
 		events.emit( 'message', message )
 	} )
+
+	socket.on( 'system.info', done => {
+		events.emit( 'system.info', data => done( data ) )
+	} )
+
 	socket.on( 'role.add', ( role, done ) => {
 		debug( 'agent joining role', role )
+		events.emit( 'role.add', role )
```
Comment (beaucollins) [inline src/agent.js:30]: What's the purpose of emitting the role?

**Comment by beaucollins** [comment]
Do we anticipate having multiple  clients connect to poll for system info?

If that's the case it might make more sense to just set up the `happychat` server to emit status info every x seconds and broadcast it to any agent that joins the specified roll.

The `role.add` client event essentially adds an agent to a Socket.IO "room" so you can broadcast to all agents in that room using something like:

```js
agent.io.to( 'role-name', { status: 'info' } )
```

This way agent clients can just connect and start listening for `system.info` without having to ask for it.

**Review by beaucollins — COMMENTED** [review]
Do we anticipate having multiple  clients connect to poll for system info?

If that's the case it might make more sense to just set up the `happychat` server to emit status info every x seconds and broadcast it to any agent that joins the specified roll.

The `role.add` client event essentially adds an agent to a Socket.IO "room" so you can broadcast to all agents in that room using something like:

``` js
agent.io.to( 'role-name', { status: 'info' } )
```

This way agent clients can just connect and start listening for `system.info` without having to ask for it.

**Comment by mjangda** [comment]
That's basically what https://github.com/Automattic/happychat/pull/121 does :)

#tinker-mind-meld

File: src/agent.js (hunk 3 above)
Comment (beaucollins) [reply src/agent.js:30]: Ok, I see that you're using this as an event to indicate that you should start the timer.

Comment (mjangda) [reply src/agent.js:30]: So we can selectively start the interval: https://github.com/Automattic/happychat/pull/121/files#diff-a42cba78a8f2c7c9988e1ba4165229bdR25

socket.io doesn't have any native events for room join/leave. Not strictly necessary as we could just have the interval running all the time but that seems like may be overkill if no one is listening for it.

**Comment by beaucollins** [comment]
Looks good to me!

I'll open issue to move the chat-list to a redux store so we can potentially just push state changes.

**Pull request merged**